# Pandas

Pandas is a python library used extensively in visual data analysis. Often, data are stored in tabular formats like the `csv`, `tsv`, `xls`. Pandas allow for the convenient loading and manipulation of these datasets either for further processing, analysis or extract meaningful data. In conjunction with other libraries like MatPlotLib, Seaborn, it provides a very powerful opportunity for visualization.

Often, pandas is aliased as pd when being loaded just to make referencing it easier:

In [1]:
import pandas as pd

## Data Structures

The main pandas data structures we will often encounter are `Series` and `DataFrame`. A Series is just a numpy array, a one-dimensional indexed array of some fixed data type. In contrast, a DataFrame is a two-dimensional data structure where each column is a Series. DataFrames are useful for representing data in a tabular format: each row corresponds to an observation and each column represents a feature of an instance.

To illustrate, let's load the country dataset:

In [2]:
countries = pd.read_csv('data/0101_Pandas_Basics/country.csv')

In [3]:
countries

country
0           Afghanistan
1               Albania
2               Algeria
3                Angola
4             Argentina
..                  ...
137             Vietnam
138  West Bank and Gaza
139         Yemen, Rep.
140              Zambia
141            Zimbabwe

[142 rows x 1 columns]

By default, pandas load the data into a DataFrame. 

In [4]:
type(countries)

pandas.core.frame.DataFrame

However, if you are trying to load data that can be presented as a single list, you can pass in the `squeeze=True` to your load method and pandas will attempt to sqeeze the data into a Series.

In [5]:
countries_series = pd.read_csv('data/0101_Pandas_Basics/country.csv', squeeze=True)

In [6]:
type(countries_series)

pandas.core.series.Series

## DataFrame Methods and Attributes

Now, let's take a look at another DataFrame and demonstrate some of the methods used in analyzing a dataset. 

In [7]:
gapminder = pd.read_csv('data/0101_Pandas_Basics/gapminder.tsv', sep='\t')

We can get the first lines of a dataset using the `head()` method which accepts an integer argument to specify how many lines to return, if this is omitted however, the method will return the first 5 lines.

In [8]:
gapminder.head()

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

Similarly, if we want to get the last lines of our dataset, we can use the `tail()` method.

In [9]:
gapminder.tail(10)

country continent  year  lifeExp       pop   gdpPercap
1694  Zimbabwe    Africa  1962   52.358   4277736  527.272182
1695  Zimbabwe    Africa  1967   53.995   4995432  569.795071
1696  Zimbabwe    Africa  1972   55.635   5861135  799.362176
1697  Zimbabwe    Africa  1977   57.674   6642107  685.587682
1698  Zimbabwe    Africa  1982   60.363   7636524  788.855041
1699  Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700  Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701  Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702  Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703  Zimbabwe    Africa  2007   43.487  12311143  469.709298

One of the most important properties of a DataFrame is its shape. It conveys the dimensionality of your data which dictates what operations can be applied to the dataset. It can be accessed using the `.shape` property.

In [10]:
gapminder.shape

(1704, 6)

We are given a tuple (1704, 6). What this means is that our data has 1704 rows and 6 columns.

For a DataFrame, there are 3 essential elements. The first one is the index, the left most column which looks like the row number. Up top is the column names. And finally the body for actual values. We can access the columns using the `.columns` property.

In [11]:
gapminder.columns

Index(['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap'], dtype='object')

This gives us what the names of the columns in array form and their datatype. You can have columns of different types but each column must have the same type. If you see object as a column, typically it means string data type but if you come from a large pipeline of data it could mean, for example another data frame or a request object or any generic python object.

In [12]:
gapminder.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
country      1704 non-null object
continent    1704 non-null object
year         1704 non-null int64
lifeExp      1704 non-null float64
pop          1704 non-null int64
gdpPercap    1704 non-null float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


Here, using the `info()` method, we can see the column names next to the datatype of the elements they contain. It also gives the the general information for our DataFrame like unique datatypes `float64, int64, object` as well as memory usage. We can also know at first glance that there are no missing data for our dataset because there are 1704 observations per column, the same number of rows we saw when we look at its shape. It is also possible to change the datatype of our column using the `astype()` method. Let's try to convert the 'year' column into int8:

In [13]:
gapminder['year'] = gapminder['year'].astype('int16')
gapminder.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
country      1704 non-null object
continent    1704 non-null object
year         1704 non-null int16
lifeExp      1704 non-null float64
pop          1704 non-null int64
gdpPercap    1704 non-null float64
dtypes: float64(2), int16(1), int64(1), object(2)
memory usage: 70.0+ KB


One of the information provided by the `info()` method is the range for the dataset indices: `RangeIndex: 1704 entries, 0 to 1703`. We can alternatively access the indices using `index()` method.

In [14]:
gapminder.index

RangeIndex(start=0, stop=1704, step=1)

The `describe()` method shows the basic statistical characteristics of each numerical feature, number of non-missing values, mean, std, range, median and quartiles

In [15]:
gapminder.describe()

year      lifeExp           pop      gdpPercap
count  1704.00000  1704.000000  1.704000e+03    1704.000000
mean   1979.50000    59.474439  2.960121e+07    7215.327081
std      17.26533    12.917107  1.061579e+08    9857.454543
min    1952.00000    23.599000  6.001100e+04     241.165877
25%    1965.75000    48.198000  2.793664e+06    1202.060309
50%    1979.50000    60.712500  7.023596e+06    3531.846989
75%    1993.25000    70.845500  1.958522e+07    9325.462346
max    2007.00000    82.603000  1.318683e+09  113523.132900

Only the columns year, lifeExp, pop and gdpPercap are displayed because they are the only columns with numerical values. In order to display the statistics on non-numeric features, we need to explicitly define which datatypes we are interested in.

In [16]:
gapminder.describe(include=['object'])

country continent
count     1704      1704
unique     142         5
top      Libya    Africa
freq        12       624

The result gave us the count, number of unique values as well as the top and occurrence frequency. Furthermore, if you want to see the count per unique value in a categorical column you can use the `value_counts()` method.

In [17]:
gapminder.country.value_counts()

Libya            12
United States    12
Kuwait           12
Congo, Rep.      12
Liberia          12
                 ..
South Africa     12
Norway           12
Denmark          12
Serbia           12
Benin            12
Name: country, Length: 142, dtype: int64

To calculate the fractions, pass `normalize=True` argument to the `value_counts()` function.

In [18]:
gapminder.country.value_counts(normalize=True)

Libya            0.007042
United States    0.007042
Kuwait           0.007042
Congo, Rep.      0.007042
Liberia          0.007042
                   ...   
South Africa     0.007042
Norway           0.007042
Denmark          0.007042
Serbia           0.007042
Benin            0.007042
Name: country, Length: 142, dtype: float64

Here, we can see that all countries in our dataset occurs exactly 12 times.

In [19]:
gapminder.country.value_counts().unique()

array([12])

## Sorting

Arguably, one of the most important methods we need is the ability to sort our data. We can do so by using the `sort_values()` method. For example, suppose we want to sort our dataset by year.

In [20]:
gapminder.head()

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

In [21]:
gapminder.sort_values(by='year', ascending=True).head()

country continent  year  lifeExp       pop    gdpPercap
0            Afghanistan      Asia  1952   28.801   8425333   779.445314
528               France    Europe  1952   67.410  42459667  7029.809327
540                Gabon    Africa  1952   37.003    420702  4293.476475
1656  West Bank and Gaza      Asia  1952   43.160   1030585  1515.592329
552               Gambia    Africa  1952   30.000    284320   485.230659

We can see that instead of the years being all jumbled up like when before we sorted our data frame, our data is sorted by year in ascending order giving us the rows with 1952 first. If however, we want to sort by multiple columns, we can pass in the array of columns as well as their corresponding sort order.

In [22]:
gapminder.sort_values(by=['year', 'country'], ascending=[True, True]).head()

country continent  year  lifeExp       pop    gdpPercap
0   Afghanistan      Asia  1952   28.801   8425333   779.445314
12      Albania    Europe  1952   55.230   1282697  1601.056136
24      Algeria    Africa  1952   43.077   9279525  2449.008185
36       Angola    Africa  1952   30.015   4232095  3520.610273
48    Argentina  Americas  1952   62.485  17876956  5911.315053

Now, not only do we have a sorted data by year, we also have it sorted by country name in ascending order.

## Indexing and Subsetting

A DataFrame can be indexed in a variety of ways. To get a single column we can use the bracket `[]` operator and pass in the name of the column.

In [23]:
gapminder['pop']

0        8425333
1        9240934
2       10267083
3       11537966
4       13079460
          ...   
1699     9216418
1700    10704340
1701    11404948
1702    11926563
1703    12311143
Name: pop, Length: 1704, dtype: int64

Boolean indexing can also be applied do a DataFrame using the syntax: `df[P(df['Name'])]` where P is some logical condition that is to be checked on the Name column. To illustrate, suppose we want to get only the subset of countries in Asis, we can do it like so:

In [24]:
gapminder[gapminder['continent'] == 'Asia']

country continent  year  lifeExp       pop    gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333   779.445314
1     Afghanistan      Asia  1957   30.332   9240934   820.853030
2     Afghanistan      Asia  1962   31.997  10267083   853.100710
3     Afghanistan      Asia  1967   34.020  11537966   836.197138
4     Afghanistan      Asia  1972   36.088  13079460   739.981106
...           ...       ...   ...      ...       ...          ...
1675  Yemen, Rep.      Asia  1987   52.922  11219340  1971.741538
1676  Yemen, Rep.      Asia  1992   55.599  13367997  1879.496673
1677  Yemen, Rep.      Asia  1997   58.020  15826497  2117.484526
1678  Yemen, Rep.      Asia  2002   60.308  18701257  2234.820827
1679  Yemen, Rep.      Asia  2007   62.698  22211743  2280.769906

[396 rows x 6 columns]

Now, we have a view of our dataset that is only limited to the continent asia with 396 distinct rows. As an aside, we can access a DataFrame column using a convenient `.name` property as long as the column name does not conflict with any DataFrame properties. So, `gapminder.continent` is equivalent to `gapminder['continent']`

In [25]:
pd.Series.equals(gapminder.continent, gapminder['continent'])

True

This wont work, however, if you have column names shape which conflicts the DataFrame property `.shape` which is reserved for the dimensionality. Moving on, if we want to subset multiple columns, instead of passing a string, we can pass in an array of column names we want to subset.

In [26]:
gapminder[['continent', 'country']]

continent      country
0         Asia  Afghanistan
1         Asia  Afghanistan
2         Asia  Afghanistan
3         Asia  Afghanistan
4         Asia  Afghanistan
...        ...          ...
1699    Africa     Zimbabwe
1700    Africa     Zimbabwe
1701    Africa     Zimbabwe
1702    Africa     Zimbabwe
1703    Africa     Zimbabwe

[1704 rows x 2 columns]

In addition, sometimes it is impractical to list all the column names you want to subset. For this, you can use the `drop()` method which accepts column names as parameter.

In [27]:
gapminder.drop(['year', 'pop'], axis='columns')

country continent  lifeExp   gdpPercap
0     Afghanistan      Asia   28.801  779.445314
1     Afghanistan      Asia   30.332  820.853030
2     Afghanistan      Asia   31.997  853.100710
3     Afghanistan      Asia   34.020  836.197138
4     Afghanistan      Asia   36.088  739.981106
...           ...       ...      ...         ...
1699     Zimbabwe    Africa   62.351  706.157306
1700     Zimbabwe    Africa   60.377  693.420786
1701     Zimbabwe    Africa   46.809  792.449960
1702     Zimbabwe    Africa   39.989  672.038623
1703     Zimbabwe    Africa   43.487  469.709298

[1704 rows x 4 columns]

So, instead of explicitly listing country, continent, lifeExp and gdpPercap the columns year and pop are excluded instead. Keep in mind however that dropping columns doesn't change the original DataFrame unless you specify the inplace parameter. This is to prevent accidental mutations to the original DataFrame.

Another useful way of accessing a group of rows and columns, there is also a `loc[]` method which accepts a single label, a list of labels or a slice object with labels, a boolean array of the same length as the DataFrame or a callable function with one argument.

In [28]:
gapminder.loc[2]

country      Afghanistan
continent           Asia
year                1962
lifeExp           31.997
pop             10267083
gdpPercap        853.101
Name: 2, dtype: object

This gives us the data for row with `index == '2'`. Note, however, that `loc[]` doesn't really look for the 3rd element but rather does a string match for the index - or for the row with index label of 2

In [29]:
gapminder.loc[2, 'country']

'Afghanistan'

This further narrows down the data to only give us the row with `index == '2'` as well as the value for the 'country' column. To refresh, our DataFrame is structured like so:

In [30]:
gapminder.head(2)

country continent  year  lifeExp      pop   gdpPercap
0  Afghanistan      Asia  1952   28.801  8425333  779.445314
1  Afghanistan      Asia  1957   30.332  9240934  820.853030

We also can pass in a slice 'country':'lifeExp' to our `loc[]` method to get all the columns in between and including country and lifeExp. 

In [31]:
gapminder.loc[2, 'country':'lifeExp']

country      Afghanistan
continent           Asia
year                1962
lifeExp           31.997
Name: 2, dtype: object

Which gives us values for country, continent, year, lifeExp.

In [32]:
gapminder.loc[5:10, 'continent':'pop']

continent  year  lifeExp       pop
5       Asia  1977   38.438  14880372
6       Asia  1982   39.854  12881816
7       Asia  1987   40.822  13867957
8       Asia  1992   41.674  16317921
9       Asia  1997   41.763  22227415
10      Asia  2002   42.129  25268405

In [33]:
gapminder.loc[:, ['year']].head()

year
0  1952
1  1957
2  1962
3  1967
4  1972

There are cases however, especially when you are unable to specify label names for the slice. In these cases, you can use `iloc[]` which accepts number parameters.

In [34]:
gapminder.iloc[5:10]

country continent  year  lifeExp       pop   gdpPercap
5  Afghanistan      Asia  1977   38.438  14880372  786.113360
6  Afghanistan      Asia  1982   39.854  12881816  978.011439
7  Afghanistan      Asia  1987   40.822  13867957  852.395945
8  Afghanistan      Asia  1992   41.674  16317921  649.341395
9  Afghanistan      Asia  1997   41.763  22227415  635.341351

As in a typical python slice, the maximum value is exclusive so we only get rows from 5 to 9. Our selection `gapminder.loc[2, 'country':'lifeExp']` is equivalent to

In [35]:
gapminder.iloc[2, 0:4]

country      Afghanistan
continent           Asia
year                1962
lifeExp           31.997
Name: 2, dtype: object

Since `loc[]` uses a match by index, it does not accept negative values and throws a `KeyError`.

In [36]:
try:
    gapminder.loc[-1]
except KeyError:
    print('KeyError')

KeyError


`iloc[]` however accepts negative indices

In [37]:
gapminder.iloc[-1]

country      Zimbabwe
continent      Africa
year             2007
lifeExp        43.487
pop          12311143
gdpPercap     469.709
Name: 1703, dtype: object

Conversely, `iloc[]` cannot accept label indices

In [38]:
try:
    subset = gapminder.iloc[:, ['year']]
except IndexError:
    print('IndexError')

IndexError


In [39]:
gapminder.loc[gapminder['country'] == 'Philippines']

country continent  year  lifeExp       pop    gdpPercap
1212  Philippines      Asia  1952   47.752  22438691  1272.880995
1213  Philippines      Asia  1957   51.334  26072194  1547.944844
1214  Philippines      Asia  1962   54.757  30325264  1649.552153
1215  Philippines      Asia  1967   56.393  35356600  1814.127430
1216  Philippines      Asia  1972   58.065  40850141  1989.374070
1217  Philippines      Asia  1977   60.060  46850962  2373.204287
1218  Philippines      Asia  1982   62.082  53456774  2603.273765
1219  Philippines      Asia  1987   64.151  60017788  2189.634995
1220  Philippines      Asia  1992   66.458  67185766  2279.324017
1221  Philippines      Asia  1997   68.564  75012988  2536.534925
1222  Philippines      Asia  2002   70.303  82995088  2650.921068
1223  Philippines      Asia  2007   71.688  91077287  3190.481016

If you just care about subsetting columns, use square brackets but if you want to subset rows and columns, use loc. Now, you can also mask data frames but this gives a warning because we pandas uses bitwise comparison

In [40]:
try:
    gapminder.loc[gapminder.country == 'Philippines' & gapminder.year == 1982]
except TypeError:
    print('TypeError')

TypeError


In [41]:
gapminder.loc[(gapminder.country == 'Philippines') & (gapminder.year == 1982)]

country continent  year  lifeExp       pop    gdpPercap
1218  Philippines      Asia  1982   62.082  53456774  2603.273765

## Applying Functions to Cells, Columns and Rows

### apply()

To apply functions to each column in a DataFrame, we can use the `apply()` method. 

In [42]:
gapminder.apply('mean')

year         1.979500e+03
lifeExp      5.947444e+01
pop          2.960121e+07
gdpPercap    7.215327e+03
dtype: float64

This gave us the mean for each column and dropped the non-numeric columns.

In [43]:
import seaborn as sns
tips = sns.load_dataset('tips')

In [44]:
tips.apply('max')

total_bill    50.81
tip              10
sex            Male
smoker          Yes
day            Thur
time          Lunch
size              6
dtype: object

### map()

The map method can be used to replace values in a column by passing a dictionary of the form `{ old_value: new_value }` as its argument

In [45]:
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

Now, suppose we want to replace the 'Yes' and 'No' values for the smoker column to `True` and `False` respectively.

In [46]:
tips.smoker = tips.smoker.map({ 'Yes': True, 'No': False })
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female  False  Sun  Dinner     2
1       10.34  1.66    Male  False  Sun  Dinner     3
2       21.01  3.50    Male  False  Sun  Dinner     3
3       23.68  3.31    Male  False  Sun  Dinner     2
4       24.59  3.61  Female  False  Sun  Dinner     4

In [47]:
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female  False  Sun  Dinner     2
1       10.34  1.66    Male  False  Sun  Dinner     3
2       21.01  3.50    Male  False  Sun  Dinner     3
3       23.68  3.31    Male  False  Sun  Dinner     2
4       24.59  3.61  Female  False  Sun  Dinner     4

The same thing can also be accomplished by using `replace()`.

In [48]:
tips.sex = tips.sex.replace({ 'Female': 'F', 'Male': 'M'})
tips.head()

total_bill   tip sex smoker  day    time  size
0       16.99  1.01   F  False  Sun  Dinner     2
1       10.34  1.66   M  False  Sun  Dinner     3
2       21.01  3.50   M  False  Sun  Dinner     3
3       23.68  3.31   M  False  Sun  Dinner     2
4       24.59  3.61   F  False  Sun  Dinner     4

## Grouping

In general, pandas grouping syntax is as follows:

`df.groupby(by=grouping_columns)[columns_to_show].function()`

1. First, the dataset is grouped by the same values for each column specified. They then become the new index in the resulting DataFrame.
2. Next, the columns of interest are filtered using the columns specified. If this is omitted, all non-groupby columns are included.
3. Lastly, functions are applied to each of the group obtained.

In [49]:
gapminder.groupby(['year']).agg('mean')

lifeExp           pop     gdpPercap
year                                       
1952  49.057620  1.695040e+07   3725.276046
1957  51.507401  1.876341e+07   4299.408345
1962  53.609249  2.042101e+07   4725.812342
1967  55.678290  2.265830e+07   5483.653047
1972  57.647386  2.518998e+07   6770.082815
1977  59.570157  2.767638e+07   7313.166421
1982  61.533197  3.020730e+07   7518.901673
1987  63.212613  3.303857e+07   7900.920218
1992  64.160338  3.599092e+07   8158.608521
1997  65.014676  3.883947e+07   9090.175363
2002  65.694923  4.145759e+07   9917.848365
2007  67.007423  4.402122e+07  11680.071820

This groups our dataset by distinct year, since there are no specific filter columns specified, all columns are included. Next the mean function is applied to all the columns but since not all columns are numeric, only those that can be applied with the mean function are displayed. In this case lifeExp, pop and gdpPercap.

In [50]:
tips.groupby(['sex', 'smoker']).agg('mean')

total_bill       tip      size
sex smoker                                
F   True     17.977879  2.931515  2.242424
    False    18.105185  2.773519  2.592593
M   True     22.284500  3.051167  2.500000
    False    19.791237  3.113402  2.711340

Group by returns a hierarchical index because it splits the dataset with the values from the first column, then the split is further split using the values from the next column, then so on.

In [51]:
for key, data in tips.groupby(['sex', 'smoker']):
    print(key)

('F', True)
('F', False)
('M', True)
('M', False)


In [52]:
tips.groupby(['sex', 'smoker']).agg('mean').reset_index()

sex smoker  total_bill       tip      size
0   F   True   17.977879  2.931515  2.242424
1   F  False   18.105185  2.773519  2.592593
2   M   True   22.284500  3.051167  2.500000
3   M  False   19.791237  3.113402  2.711340

## DataFrame Transformations

There are different ways in adding columns to a DataFrame. One of which is by using the `insert()` method. Suppose we want to calculate the tip percentage from the total bill and add them as a new column, we can do that like this:

In [53]:
tip_percentage = tips.tip / tips.total_bill
tips.insert(loc=len(tips.columns), column='tip_percentage', value=tip_percentage)
tips.head()

total_bill   tip sex smoker  day    time  size  tip_percentage
0       16.99  1.01   F  False  Sun  Dinner     2        0.059447
1       10.34  1.66   M  False  Sun  Dinner     3        0.160542
2       21.01  3.50   M  False  Sun  Dinner     3        0.166587
3       23.68  3.31   M  False  Sun  Dinner     2        0.139780
4       24.59  3.61   F  False  Sun  Dinner     4        0.146808

It is possible too to add without an intermediate Series instance by calling the column directly. Suppose we want to divide the total_bill by the size of the meal.

In [54]:
tips['bill_to_size_ratio'] = tips['total_bill'] / tips['size']
tips.head()

total_bill   tip sex smoker  day    time  size  tip_percentage  \
0       16.99  1.01   F  False  Sun  Dinner     2        0.059447   
1       10.34  1.66   M  False  Sun  Dinner     3        0.160542   
2       21.01  3.50   M  False  Sun  Dinner     3        0.166587   
3       23.68  3.31   M  False  Sun  Dinner     2        0.139780   
4       24.59  3.61   F  False  Sun  Dinner     4        0.146808   

   bill_to_size_ratio  
0            8.495000  
1            3.446667  
2            7.003333  
3           11.840000  
4            6.147500